huaweicloud/services/antiddos/resource_huaweicloud_antiddos_basic.go

```diff
@@ -138,27 +145,90 @@ func resourceCloudNativeAntiDdosUpdate(ctx context.Context, d *schema.ResourceDa
 	region := cfg.GetRegion(d)
 	client, err := cfg.AntiDDosV1Client(region)
 	if err != nil {
-		return diag.Errorf("error creating antiddos client: %s", err)
+		return diag.Errorf("error creating AntiDDoS v1 client: %s", err)
+	}
+
+	warnAlertClient, err := cfg.AntiDDosV2Client(region)
+	if err != nil {
+		return diag.Errorf("error creating AntiDDoS v2 client: %s", err)
 	}
 
 	eipID := d.Get("eip_id").(string)
-	thresholdID := getTrafficThresholdID(d.Get("traffic_threshold").(int))
+	if d.HasChange("traffic_threshold") {
+		thresholdID := getTrafficThresholdID(d.Get("traffic_threshold").(int))
 
-	if err := updateAntiDdosTrafficThreshold(ctx, d, client, eipID, thresholdID, false); err != nil {
-		return diag.FromErr(err)
+		if err := updateAntiDdosTrafficThreshold(ctx, d, client, eipID, thresholdID, false); err != nil {
+			return diag.FromErr(err)
+		}
+	}
+
+	if d.HasChange("topic_urn") {
+		topicUrn := d.Get("topic_urn").(string)
+		displayName := getDisplayName(topicUrn)
+
+		if err := updateAntiDdosWarnAlert(warnAlertClient, topicUrn, displayName); err != nil {
+			return diag.FromErr(err)
+		}
 	}
 
 	d.SetId(eipID)
 	return resourceCloudNativeAntiDdosRead(ctx, d, meta)
 }
 
+func updateAntiDdosWarnAlert(client *golangsdk.ServiceClient, topicUrn string, displayName string) error {
+	var updateOpts warnalertsdk.UpdateOpsBuilder
+	if topicUrn != "" {
+		updateOpts = warnalertsdk.UpdateOps{
+			TopicUrn: topicUrn,
+			WarnConfig: &warnalertsdk.WarnConfig{
+				EnableAntiDDoS: utils.Bool(true),
+			},
+			DisplayName: displayName,
+		}
+	} else {
+		alarmResult, err := warnalertsdk.GetWarnAlert(client).Extract()
+		if err != nil {
+			return fmt.Errorf("error retrieving AntiDDoS alarm configuration: %s", err)
+		}
+		updateOpts = warnalertsdk.UpdateOps{
+			TopicUrn: alarmResult.TopicUrn,
+			WarnConfig: &warnalertsdk.WarnConfig{
+				EnableAntiDDoS: utils.Bool(false),
+			},
+			DisplayName: alarmResult.DisplayName,
+		}
+	}
+
+	if _, err := warnalertsdk.UpdateWarnAlert(client, updateOpts).Extract(); err != nil {
+		return fmt.Errorf("error updating AntiDDoS alarm configuration: %s", err)
+	}
+
+	return nil
+}
+
+// "topic_urn":"urn:smn:cn-south-1:09f960944c80f4802f85c333e0ed1d98:tf_test"
+func getDisplayName(topUrn string) string {
```

Comment:
it's better to update the function name
```suggestion
func getSmnDisplayName(topUrn string) string {
```